### ISSUE TITLE
Array Curve container

### ISSUE THREAD

**Issue by heinezen** [SFTtech/openage#1678]
We should add a `curve::Array` type to our curve container classes to record changes to a fixed size sequence over time. In comparison to https://github.com/SFTtech/openage/issues/1677 we can probably optimize a lot more since the size doesn't change over time (i.e. no inserts or erasures).

The new array type should implement these methods:

**Read**

| Method          | Description                                                 |
| --------------- | ----------------------------------------------------------- |
| `get(t, i)`        | Get value of index `i` at time `t`                        |
| `get(t)`        | Get array at time `t`                        |
| `size()`        | Get size of the array                        |
| `frame(t, i)`      | Get the previous keyframe (time and value) of index `i` before or at `t` |
| `next_frame(t, i)` | Get the next keyframe (time and value) of index `i` after `t`            |

There should also be an iterator type to iterate over the array at time `t`.

**Modify**

| Method                  | Description                                                                       |
| ----------------------- | --------------------------------------------------------------------------------- |
| `set_insert(t, i, value)`  | Insert a new keyframe value at time `t` at index `i`                                           |
| `set_last(t, i, value)`    | Insert a new keyframe value at time `t`; delete all keyframes after time `t` at index `i`      |
| `set_replace(t, i, value)` | Insert a new keyframe value at time `t` at index `i`; remove all other keyframes with time `t` at index `i` |

**Copy**

| Method           | Description                                                                                      |
| ---------------- | ------------------------------------------------------------------------------------------------ |
| `sync(Curve, t)` | Replace all keyframes from self after time `t` with keyframes from source `Curve` after time `t` |

**Comment by jere8184** [comment]
@heinezen can I take this

**Comment by heinezen** [comment]
Sure, I think there's currently nothing blocking this feature.

**Comment by jere8184** [comment]
So will this be an array of vectors of Keyframes, essentially?

**Comment by heinezen** [comment]
Yeah that's basically it.

**Comment by jere8184** [comment]
Should I use something other than KeyframeContainers, queue and map don't seem to use it.

**Comment by heinezen** [comment]
For the array curve, `KeyframeContainer`s are fine because the size of the array is fixed. For the other containers, elements are not guaranteed to be persistent. They can get erased and later added again, which has to be accounted for (that's why there's this dead/alive logic for their keyframes).

**Comment by jere8184** [comment]
So for the iterator, do we iterate over keyframes with the closest time less than or equal to t?

**Comment by heinezen** [comment]
Yes, when iterating over the array at time `t`, it should iterate over the elements that are valid at time `t`

So the result should be the same as calling all the indices individually for the same time: `get(t, 0)` ... `get(t, n)`

**Comment by jere8184** [comment]
No erase method?

**Comment by heinezen** [comment]
The array would have a fixed size from the start for the duration of its lifetime, so removing elements wouldn't make sense.

**Comment by jere8184** [comment]
What about removing elements from keyframecontainers?

**Comment by heinezen** [comment]
Is there a specific example where removing keyframes would be necessary? It's not used by any other curve. If we need it in the future, we could always add it, but I don't think it belongs to the public API of a curve array.

**Comment by jere8184** [comment]
`BaseCurve` has an `erase `method, which is public.

**Comment by heinezen** [comment]
@jere8184 Hmm... I think we could add it then. Certainly wouldn't hurt!

**Comment by jere8184** [comment]
How should the iterator handle KeyframeContainers that contain multiple keyframes for a given time?

**Comment by heinezen** [comment]
@jere8184 In that case, the last frame that was inserted (as in: by insertion order) for that time should be used.

**Comment by jere8184** [comment]
So I've implemented the array curve using `KeyframeContainer`s this differs from the other Container curves (`queue `and `map`) as they store their elements in `element_wrapper`s this difference was supposedly because we will not erase `Keyframe`s from the `array` right?

> For the array curve, `KeyframeContainer`s are fine because the size of the array is fixed, so all elements exist with a value over the duration of the array's existance. For the other containers, elements are not guaranteed to always be persistent. They can get erased and later added again, which has to be accounted for (that's why there's this dead/alive logic for their keyframes).

I'm just making sure I understand correctly before updating the curve documentation file.

Where does the idea that we shouldn't erase `keyframes` from `array` come from?

**Comment by heinezen** [comment]
> Where does the idea that we shouldn't erase keyframes from array come from?

I think that's more a misunderstanding. My point was that *elements* shouldn't be erased for arrays (in comparison to maps/quues), since the array always has a fixed size. Removing *keyframes* from specific containers is okay, it's just not a very common operation.
